doc/threshold/SONiC Threshold feature spec.md

```diff
@@ -0,0 +1,589 @@
+
+## 2.2 Functional Description
+
+In a live network, tracking and monitoring buffer usage on a switch can be extremely useful in detecting and isolating potential and actual congestion events in the network. The availability of telemetry information w.r.t buffer usage on network nodes can help to develop network-level monitoring applications. 
+
+The watermark feature in SONiC makes the buffer usage counters readily available in COUNTERS\_DB for multiple users. The watermarks can be used to track peak buffer occupancy across users and intervals. In addition to this, it may also be useful for the user to know when the buffer usage crosses a certain threshold. The threshold feature implements this specific functionality.
+
+The threshold feature allows configuration of the threshold value as a percentage of the buffer limits(MMU configuration via buffer profile or otherwise). The SAI APIs take threshold as a percentage and convert this percentage to an absolute value based upon the appropriate buffer limits configured. For example, if the configured buffer limit for a given priority-group of a port (shared) buffer is 5000 bytes, a threshold value of 10% configured by the user translates to an absolute value of 500 bytes. SAI configures the threshold for the buffer  as 500 bytes and a threshold breach event gets generated if the buffer usage crosses 500 bytes. Note that the original buffer limit and the absolute threshold value configured is not exposed to the user, the user can simply set a threshold in terms of % and be notified when the threshold is crossed.
+
+The supported buffers for threshold are:
+
+1. Ingress port priority group
+	1. Shared - Specifies the shared buffer usage on a priority group.
+	2. Headroom - Specifies the additional buffer limit that can be utilized when the shared limits are exhausted in the case where flow control is enabled. It accounts for the round trip time of data to be received once flow control is asserted. 
+2. Egress port queues
+	1. Unicast 
+	2. Multicast
+ 
+
+# 3 Design
+## 3.1 Overview
+
+![Threshold architecture](images/Threshold_arch.jpg)
+
+The above diagram illustrates the architecture of the threshold feature within SONiC. The call flow sequence is listed below:
+
+1. CLI takes threshold configuration and populates it into THRESHOLD\_TABLE of config DB.
+2. The ThresholdOrch component of Orchagent in SWSS docker picks up the config DB updates on THRESHOLD\_TABLE and converts the config into SAIREDIS calls. ASIC\_DB is updated by SAIREDIS and SYNCD gets notified of incoming updates.
+3. SYNCD invokes the SAI APIs for configuring the threshold. 
+4. SAI configures the ASIC with threshold configuration.
+5. ASIC notifies SAI of a threshold breach notification.
+6. SAI collates the breach information into a protobuf message and sends it out to the configured collector (in this case, ThresholdMgr) on a socket.

```

Comment:
Need some clarification -- is SAI sending the protobuf message? Since generating the protobuf message is device independent, can SAI just collate the information and let the application (syncd) pack this information into the protobuf message and send it out.
 For instance, on a breach notification SAI can invoke something like SAI_SWITCH_ATTR_TAM_EVENT_NOTIFY and let the application update COUNTERS\_DB.

Reply:
Hi,

Yes, SAI is sending the protobuf message out.
The breach protobuf sent by SAI contains buffer usage stats for all front panel port's PGs and queues, buffer-pools etc at the instant of the breach in addition to the general breach information.  The stats collected at the breach instant are packed into a protobuf and sent out to the collector. 

SAI TAM spec already specifies support for sending this breach data to a collector. The same is used on SONIC.